homeassistant/components/wled/light.py

```diff
@@ -0,0 +1,218 @@
+        except WLEDError:
+            _LOGGER.error("An error occurred while turning off WLED light.")
+            self._available = False
+        self.async_schedule_update_ha_state()
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on the light."""
+        data = {ATTR_ON: True, ATTR_SEGMENT_ID: self._segment}
+
+        if ATTR_COLOR_TEMP in kwargs:
+            mireds = color_util.color_temperature_kelvin_to_mired(
+                kwargs[ATTR_COLOR_TEMP]
+            )
+            data[ATTR_COLOR_PRIMARY] = tuple(
+                map(int, color_util.color_temperature_to_rgb(mireds))
+            )
+
+        if ATTR_HS_COLOR in kwargs:
+            hue, sat = kwargs[ATTR_HS_COLOR]
+            data[ATTR_COLOR_PRIMARY] = color_util.color_hsv_to_RGB(hue, sat, 100)
+
+        if ATTR_TRANSITION in kwargs:
+            data[ATTR_TRANSITION] = kwargs[ATTR_TRANSITION]
+
+        if ATTR_BRIGHTNESS in kwargs:
+            data[ATTR_BRIGHTNESS] = kwargs[ATTR_BRIGHTNESS]
+
+        if ATTR_EFFECT in kwargs:
+            data[ATTR_EFFECT] = kwargs[ATTR_EFFECT]
+
+        # Support for RGBW strips
+        if self._rgbw and any(x in (ATTR_COLOR_TEMP, ATTR_HS_COLOR) for x in kwargs):
+            data[ATTR_COLOR_PRIMARY] = color_util.color_rgb_to_rgbw(
+                *data[ATTR_COLOR_PRIMARY]
+            )
+
+        try:
+            await self.wled.light(**data)
+
+            self._state = True
+
+            if ATTR_BRIGHTNESS in kwargs:
+                self._brightness = kwargs[ATTR_BRIGHTNESS]
+
+            if ATTR_EFFECT in kwargs:
+                self._effect = kwargs[ATTR_EFFECT]
+
+            if ATTR_HS_COLOR in kwargs:
+                self._color = kwargs[ATTR_HS_COLOR]
+
+            if ATTR_COLOR_TEMP in kwargs:
+                self._color = color_util.color_temperature_to_hs(mireds)
+
+        except WLEDError:
+            _LOGGER.error("An error occurred while turning on WLED light.")
+            self._available = False
+        self.async_schedule_update_ha_state()
+
+    async def _wled_update(self) -> None:
+        """Update WLED entity."""
+        self._brightness = self.wled.device.state.brightness
+        self._effect = self.wled.device.state.segments[self._segment].effect.name
+        self._state = self.wled.device.state.on
+
+        color = self.wled.device.state.segments[self._segment].color_primary
+        if self._rgbw:
+            color = color_util.color_rgbw_to_rgb(*color)
+        self._color = color_util.color_RGB_to_hs(*color)
+
+        playlist = self.wled.device.state.playlist
+        if playlist == -1:
+            playlist = None
+
+        preset = self.wled.device.state.preset
+        if preset == -1:
+            preset = None
+
+        self._attributes = {
+            ATTR_INTENSITY: self.wled.device.state.segments[self._segment].intensity,
+            ATTR_PALETTE: self.wled.device.state.segments[self._segment].palette.name,
```

Comment:
You are dereferencing `self.wled.device` (and `self.wled.device.state`) in several parts of the file. Maybe worth considering having a `self._device` and/or `self._state` set in the init and use that?

Reply:
Not sure how that makes a difference in this case?
It would add extra variables that refer to the same object.

Reply:
Oh btw, the `self.wled.device` is immutable, which is the reason why some things are stored in a local class variable.